### PR TITLE
Feature/candidate form

### PR DESCRIPTION
**Pull request by aryanjain28**
**Candidate form component**
(Sub component for rules page)

1. Validations using YUP.
2. UI changes using CSS file (basic animations).
3. API post call to the backend for saving the candidate details (File- candidateFormApi.js). 
4. Implemented redux-saga.
5. Used reactstrap for all UI components - (Functionalities like - Spinner, Toast, Formfeedback, Card, Input).
6. Unit testing - (Reducer and Saga).

Files created: 
         1. candidateFormActions, 
         2. candidateFormConstants, 
         3. candidateFormReducer, 
         4. rootSagas
         5. UserProfileComponent/index
         6. UserProfileComponent/infoComponent.css         
         7. UserProfileContainers.jsx
         8. candidateFormApi
         9. testFiles

### CODE REVIEW

File: src/containers/UserProfileContainer.jsx [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import UserProfileComponent from "components/UserProfileComponent"
+import { candidateFormRequestAction } from "actions/candidateFormActions";
+
+const schema = yup.object().shape({
+    fName: yup.string().required().matches(/^[A-Za-z '.-]*$/, 'fName should be valid'),
+    lName: yup.string().required().matches(/^[A-Za-z '.-]*$/, 'lName should be valid'),
+    mobile: yup.string().required().matches(/^[0-9]{10}$/, 'mobile should be a 10-digit valid numeric value')
+});
+
+const reducer = (state, action) =>{
```
Comment (Rutu2k) [inline src/containers/UserProfileContainer.jsx:14]: you can also shift your reducer from container to plain js file

File: src/reducers/__tests__/candidateFormReducer.test.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    candidateFormReducer,
+    initialState
+} from "reducers/candidateFormReducer";
+
+import {CANDIDATE_FORM_ACTIONS} from "constants/candidateFormConstants";
+import { 
+    candidateFormRequestAction, 
+    candidateFormSuccessAction,
+    candidateFormFailureAction
+ } from "actions/candidateFormActions"
```
Comment (Rutu2k) [inline src/reducers/__tests__/candidateFormReducer.test.js:11]: you can use 
describe("test suit", () => {
    it("test case1", ()=> {
 })
})

File: src/sagas/__tests__/rootSagas.test.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        expect(JSON.stringify(value1)).toBe(JSON.stringify(value2));
+
+        const value3 = gen.next(data).value
+        const value4 = put(candidateFormFailureAction(data));
+
+        expect(JSON.stringify(value3)).toBe(JSON.stringify(value4));
+
+    })
+
+    // it("Must dispatch success action", () => {
```
Comment (Rutu2k) [inline src/sagas/__tests__/rootSagas.test.js:47]: remove commented code

File: src/sagas/__tests__/rootSagas.test.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    it("API call for REQUEST", () => {
+        const value1 = gen.next().value
+        const value2 = call(candidateInfoPostApi, action.payload)
+        
+        expect(JSON.stringify(value1)).toBe(JSON.stringify(value2));
+    })
+
+    it("FAILURE action", () => {
+
+        const value1 = gen.next().value
```
Comment (Rutu2k) [inline src/sagas/__tests__/rootSagas.test.js:35]: you should use meaningful variable names

File: src/sagas/rootSagas.js [hunk 5]
```diff
@@ -1,5 +1,30 @@
-import { all } from "redux-saga/effects";
+import candidateInfoPostApi from "apis/candidateFormApi";
+import { CANDIDATE_FORM_ACTIONS } from "constants/candidateFormConstants";
+import { all, call, put, takeLatest } from "redux-saga/effects";
```
Comment (Rutu2k) [inline src/sagas/rootSagas.js:3]: rearrange the import statements, imports from libraries should be on top.

File: src/sagas/rootSagas.js [hunk 6]
```diff
@@ -1,5 +1,30 @@
...
+    }
+}
+
+//watcher saga
+export function* userSaga() {
+  yield takeLatest(CANDIDATE_FORM_ACTIONS.CANDIDATE_REQUEST, candidateFormSaga);
+}
 
 export default function* rootSaga() {
-  yield all([]);
```
Comment (Rutu2k) [inline src/sagas/rootSagas.js:4]: why two all[] ? , and  you should make different file for your saga, and import that in root saga.

File: src/store.js [hunk 7]
```diff
@@ -1,11 +1,12 @@
...
 import createSagaMiddleware from "redux-saga";
 
-import rootReducer from "reducers/rootReducer";
+// import rootReducer from "reducers/rootReducer";
 import rootSagas from "sagas/rootSagas";
+import { candidateFormReducer } from "reducers/candidateFormReducer";
 
 const sagaMiddleware = createSagaMiddleware();
 
-export const store = createStore(rootReducer, applyMiddleware(sagaMiddleware));
```
Comment (Rutu2k) [inline src/store.js:9]: right now there is no need of rootReducer here

**Review by Rutu2k — COMMENTED** [review]
Please look into the comments, and ping me if you find any difficulty.

File: src/constants/candidateFormConstants.js [hunk 8]
```diff
@@ -0,0 +1,5 @@
+export const CANDIDATE_FORM_ACTIONS = {
```
Comment (shailendrakanherkar18) [inline src/constants/candidateFormConstants.js:1]: We can keep action keys same for e.g 

GET_ACTION/POST_ACTION...
SUCESS_ACTION
FAILURE_ACTION

File: src/components/ProfileComponent/index.jsx [hunk 9]
```diff
@@ -0,0 +1,20 @@
+import { Container, Row, Col } from "reactstrap";
+import UserProfileContainer from "containers/UserProfileContainer"
+import RulesContainer from "containers/RulesContainer"
+
+const ProfileComponent = () => {
+    return(
+        <Container className="no-gutters d-flex my-height" fluid = {true}>
```
Comment (shailendrakanherkar18) [inline src/components/ProfileComponent/index.jsx:7]: <Container className="no-gutters d-flex my-height" fluid>
        boolean suceess flag can be direct passed

File: src/components/UserProfileComponent/index.jsx [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+    Label,
+    Card, 
+    CardBody, 
+    Button,
+    Row, 
+    Col, 
+    Spinner,
+    Toast,
+    ToastHeader
+} from 'reactstrap';
```
Comment (shailendrakanherkar18) [inline src/components/UserProfileComponent/index.jsx:15]: Are we not following reactstrap wrapper ??

File: src/components/UserProfileComponent/index.jsx [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+import "./infoComponent.css"
+
+const UserProfileComponent = (props) => {
```
Comment (shailendrakanherkar18) [inline src/components/UserProfileComponent/index.jsx:18]: destruct prop before using

File: src/components/ProfileComponent/index.jsx [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+        <Row>
+          <Col lg={7} >
+              <UserProfileContainer/>
+          </Col>
+          <Col lg={5} className = "bg-success">
```
Comment (Suhas010) [inline src/components/ProfileComponent/index.jsx:12]: don't need to add extra space around =

File: src/components/UserProfileComponent/index.jsx [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+                    <h2 className="pb-3 text-success text-center font-weight-bolder">OnCot</h2>
+                    <h5 className="pb-3 text-center font-weight-bolder">Please fill below details</h5>
+                    
+                    <Form>
+                        <Row className="py-3">
+                            <Col>
+                                <FormGroup>
+                                    <Input 
+                                        className="shadow"
+                                        onChange={props.fNameChanged}
```
Comment (Suhas010) [inline src/components/UserProfileComponent/index.jsx:34]: handleNameChange would be more readable here

File: src/components/UserProfileComponent/index.jsx [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+                    <h5 className="pb-3 text-center font-weight-bolder">Please fill below details</h5>
+                    
+                    <Form>
+                        <Row className="py-3">
+                            <Col>
+                                <FormGroup>
+                                    <Input 
+                                        className="shadow"
+                                        onChange={props.fNameChanged} 
+                                        invalid={props.fNameInvalid.invalid}
```
Comment (Suhas010) [inline src/components/UserProfileComponent/index.jsx:35]: destructure and Invalid is repeating here can avoid from fName,
fName => firstName

File: src/containers/UserProfileContainer.jsx [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+        default: return state;
+    }
+}
+
+const UserProfileContainer = () => {
+
+    const dispatch = useDispatch();
+    const globalState = useSelector(state => state);
+
+    console.log("globalState: ", globalState);
```
Comment (Suhas010) [inline src/containers/UserProfileContainer.jsx:44]: remove log

File: src/containers/UserProfileContainer.jsx [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+    const initialUserState = {
+        fName: { value: "", invalidState: {invalid:false, message:""} },
+        lName: { value: "", invalidState: {invalid:false, message:""} },
```
Comment (Suhas010) [inline src/containers/UserProfileContainer.jsx:48]: can be renamed to state: {valid: true/false}
it seems more readable

File: src/containers/UserProfileContainer.jsx [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+            const error = globalState.error;
+
+            dispatch(candidateFormRequestAction(data));
+        })
+        .catch((error) => {
+            const errors = error.errors;
+
+            var hint = ""
+            errors.forEach(error => {
+                hint = error.split(" ")[0]
```
Comment (Suhas010) [inline src/containers/UserProfileContainer.jsx:94]: here we have scope to optimize this catch block. please give a second thought here

File: src/containers/UserProfileContainer.jsx [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+                if (hint === "fName"){
+                    const message = error.replace("fName", "First name");
+                    setUserState({type:"fNameInvalid", payload:{value: fName, invalidState:{invalid:true, message:message}}});
+                }
+                if (hint === "lName"){
+                    const message = error.replace("lName", "Last name")
+                    setUserState({type:"lNameInvalid", payload:{value: lName, invalidState:{invalid:true, message:message}}});
```
Comment (Suhas010) [inline src/containers/UserProfileContainer.jsx:101]: code formatting is missing?

File: src/components/ProfileComponent/index.jsx [hunk 19]
```diff
@@ -0,0 +1,20 @@
+import { Container, Row, Col } from "reactstrap";
```
[the remark this answers is not in the document]
Comment (shailendrakanherkar18) [reply src/components/ProfileComponent/index.jsx:1]: Sync branch with eslint changes of vaibhav

File: src/components/ProfileComponent/index.jsx (hunk 9 above)
Comment (aryanjain28) [reply src/components/ProfileComponent/index.jsx:7]: @Rutu2k Please have a look here.

File: src/containers/UserProfileContainer.jsx [hunk 20]
```diff
@@ -0,0 +1,131 @@
...
+
+    const handleSubmit = (event) => {
+        event.preventDefault()
+
+        const fName = userState.fName.value;
+        const lName = userState.lName.value;
+        const mobile = userState.mobile.value;
+
+        schema.validate({
+            fName: fName,
```
Comment (vaibhavchalse99) [inline src/containers/UserProfileContainer.jsx:78]: use shorthand here

File: src/containers/UserProfileContainer.jsx [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+            lName: lName,
+            mobile: mobile
+        }, {abortEarly:false})
+        .then(() => {
+
```
Comment (vaibhavchalse99) [inline src/containers/UserProfileContainer.jsx:83]: why blank space ?
remove it

File: src/containers/UserProfileContainer.jsx [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+            const data = {fName, lName, mobile};
+            const error = globalState.error;
+
+            dispatch(candidateFormRequestAction(data));
+        })
+        .catch((error) => {
+            const errors = error.errors;
+
+            var hint = ""
```
Comment (vaibhavchalse99) [inline src/containers/UserProfileContainer.jsx:92]: use let instead of var

File: src/reducers/__tests__/candidateFormReducer.test.js [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+        lName : action.payload.lName,
+        email : action.payload.email,
+        mobile : action.payload.mobile
+    };
+    expect(JSON.stringify(expectedSuccess)).toBe(JSON.stringify(recievedSuccess));
+});
+
+
+
+test("candidateFormReducer Failure", () => {
```
Comment (vaibhavchalse99) [inline src/reducers/__tests__/candidateFormReducer.test.js:42]: why too much blank spaces?

File: src/reducers/candidateFormReducer.js [hunk 24]
```diff
@@ -0,0 +1,42 @@
+import { CANDIDATE_FORM_ACTIONS } from "constants/candidateFormConstants";
+
+export const initialState = {
+    loading : false,
+    error : false,
+
```
Comment (vaibhavchalse99) [inline src/reducers/candidateFormReducer.js:6]: please remove unnecessary blank spaces from all files

File: src/reducers/candidateFormReducer.js [hunk 25]
```diff
@@ -0,0 +1,42 @@
...
+    fName : undefined,
+    lName : undefined,
+    mobile: undefined,
+    errorMsg : undefined
+}
```
Comment (vaibhavchalse99) [inline src/reducers/candidateFormReducer.js:11]: use one data field and move all candidate related data in it
eg  = 
data = {
firstname : ' ',
......
}

File: src/reducers/candidateFormReducer.js [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+
+export const candidateFormReducer = (state = initialState, action) => {
```
Comment (vaibhavchalse99) [inline src/reducers/candidateFormReducer.js:13]: use immer library here

File: src/sagas/__tests__/rootSagas.test.js [hunk 27]
```diff
@@ -0,0 +1,45 @@
+import { 
+    candidateFormRequestAction, 
+    candidateFormSuccessAction, 
+    candidateFormFailureAction 
+} from "actions/candidateFormActions";
+import candidateInfoPostApi from "apis/candidateFormApi";
+import { call, put } from "redux-saga/effects";
```
Comment (vaibhavchalse99) [inline src/sagas/__tests__/rootSagas.test.js:7]: third paty libraries should be imported 1st

File: public/index.html [hunk 28]
```diff
@@ -24,9 +24,9 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
-    <title>React App</title>
+    <title>OnCot</title>
   </head>
-  <body>
+  <body style="background-color:  #000000;">
```
Comment (shailendrakanherkar18) [inline public/index.html:29]: Please apply background color via css.

Comment (shailendrakanherkar18) [reply public/index.html:29]: only dynamic css should apply via style

File: src/components/UserProfileComponent/index.jsx [hunk 29]
```diff
@@ -0,0 +1,144 @@
...
+
+  firstNameIsValid: PropTypes.objectOf(PropTypes.object).isRequired,
+  lastNameIsValid: PropTypes.objectOf(PropTypes.object).isRequired,
+  mobileIsValid: PropTypes.objectOf(PropTypes.object).isRequired,
+
+  result: PropTypes.objectOf(PropTypes.object).isRequired,
+  showToast: PropTypes.bool.isRequired,
+};
+
+export default UserProfileComponent;
```
Comment (shailendrakanherkar18) [inline src/components/UserProfileComponent/index.jsx:144]: Wrap with React.memo

File: src/containers/UserProfileContainer.jsx [hunk 30]
```diff
@@ -0,0 +1,146 @@
+import { useReducer, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import * as yup from 'yup';
+import produce from 'immer';
+
+import UserProfileComponent from 'components/UserProfileComponent';
+import { candidateFormRequestAction } from 'actions/candidateFormActions';
+
+const schema = yup.object().shape({
```
Comment (shailendrakanherkar18) [inline src/containers/UserProfileContainer.jsx:9]: can we move to separate file please, anything related to plain js can be maintian in separate file

File: src/containers/UserProfileContainer.jsx [hunk 31]
```diff
@@ -0,0 +1,146 @@
...
+  lName: yup.string()
+    .required('Last name is a required field')
+    .matches(/^[A-Za-z '.-]*$/, 'Last name should be valid'),
+
+  mobile: yup.string()
+    .max(10, 'Mobile should be a valid 10-digit numeric value')
+    .min(10, 'Mobile should be a valid 10-digit numeric value'),
+});
+
+const reducer = (state, action) => {
```
Comment (shailendrakanherkar18) [inline src/containers/UserProfileContainer.jsx:23]: can we move to separate file please, anything related to plain js can be maintian in separate file

File: src/containers/UserProfileContainer.jsx [hunk 32]
```diff
@@ -0,0 +1,146 @@
...
+  const initialUserState = {
+    fName: { value: '', state: { valid:true, message:'' } },
+    lName: { value: '', state: { valid:true, message:'' } },
+    mobile: { value: '', state: { valid:true, message:'' } },
+  };
+
+  const [userState, setUserState] = useReducer(reducer, initialUserState);
+  const [showToast, setShowToast] = useState(false);
+
+  const handleFirstNameChange = (event) => {
```
Comment (shailendrakanherkar18) [inline src/containers/UserProfileContainer.jsx:69]: wrap all memeber function with useCallback.

**Pull request merged**